File: examples/javascript/embeddings.mjs
#!/usr/bin/env node
// Create text embeddings through RekAI using the built-in fetch (Node 18+).
//
// Usage:
//   node javascript/embeddings.mjs "first text" "second text" ...
//
// Environment:
//   REKAI_API_URL        API base URL (default http://localhost:8000)
//   MODEL                embeddings model (default "echo", keyless)
//   REKAI_PROVIDER_KEY   optional BYOK key, sent as X-Provider-Key
//   REKAI_GATEWAY_KEY    optional gateway key, sent as Authorization: Bearer
//                        (only needed if the deployment has REKAI_API_KEYS set)

const API_URL = process.env.REKAI_API_URL || "http://localhost:8000";
const MODEL = process.env.MODEL || "echo";

async function embed(inputs) {
  const headers = { "Content-Type": "application/json" };
  if (process.env.REKAI_PROVIDER_KEY) {
    headers["X-Provider-Key"] = process.env.REKAI_PROVIDER_KEY;
  }
  if (process.env.REKAI_GATEWAY_KEY) {
    headers["Authorization"] = `Bearer ${process.env.REKAI_GATEWAY_KEY}`;
  }

  let res;
  try {
    res = await fetch(`${API_URL}/v1/embeddings`, {
      method: "POST",
      headers,
      body: JSON.stringify({ model: MODEL, input: inputs }),
    });
  } catch (err) {
    throw new Error(`Could not reach RekAI at ${API_URL}: ${err.message}`);
  }

  if (!res.ok) {
    throw new Error(`API error ${res.status}: ${await res.text()}`);
  }
  return res.json();
}

function cosine(a, b) {
  let dot = 0;
  let na = 0;
  let nb = 0;
  for (let i = 0; i < a.length; i++) {
    dot += a[i] * b[i];
    na += a[i] * a[i];
    nb += b[i] * b[i];
  }
  return na && nb ? dot / (Math.sqrt(na) * Math.sqrt(nb)) : 0;
}

const texts =
  process.argv.length > 2
    ? process.argv.slice(2)
    : ["a cat sat on the mat", "a feline rested on the rug", "quarterly revenue grew"];

try {
  const result = await embed(texts);
  const vectors = result.embeddings;
  console.log(`provider=${result.provider} model=${result.model} dim=${vectors[0].length}`);
  for (let i = 1; i < vectors.length; i++) {
    const sim = cosine(vectors[0], vectors[i]).toFixed(4);
    console.log(`  sim(${JSON.stringify(texts[0])}, ${JSON.stringify(texts[i])}) = ${sim}`);
  }
} catch (err) {
  console.error(err.message);
  process.exit(1);
}

File: examples/python/embeddings.py
#!/usr/bin/env python3
"""Create text embeddings through RekAI, using only the stdlib.

The keyless `echo` provider returns deterministic vectors, so this runs out of
the box:

    REKAI_API_URL=http://localhost:8000 python python/embeddings.py

For real vectors point at an embeddings model + provider key, e.g.

    MODEL=text-embedding-3-small REKAI_PROVIDER_KEY=sk-... \
    python python/embeddings.py

(Ollama works keyless too: MODEL=nomic-embed-text with Ollama running.)
"""

from __future__ import annotations

import json
import math
import os
import urllib.request

API_URL = os.environ.get("REKAI_API_URL", "http://localhost:8000")
MODEL = os.environ.get("MODEL", "echo")


def embed(inputs: list[str]) -> dict:
    body = {"model": MODEL, "input": inputs}
    headers = {"Content-Type": "application/json"}
    key = os.environ.get("REKAI_PROVIDER_KEY")
    if key:
        headers["X-Provider-Key"] = key
    gateway_key = os.environ.get("REKAI_GATEWAY_KEY")
    if gateway_key:
        headers["Authorization"] = f"Bearer {gateway_key}"
    req = urllib.request.Request(
        f"{API_URL}/v1/embeddings", data=json.dumps(body).encode(), headers=headers
    )
    with urllib.request.urlopen(req) as resp:
        return json.load(resp)


def cosine(a: list[float], b: list[float]) -> float:
    dot = sum(x * y for x, y in zip(a, b))
    na = math.sqrt(sum(x * x for x in a))
    nb = math.sqrt(sum(y * y for y in b))
    return dot / (na * nb) if na and nb else 0.0


def main() -> None:
    texts = ["a cat sat on the mat", "a feline rested on the rug", "quarterly revenue grew"]
    result = embed(texts)
    vectors = result["embeddings"]
    print(f"provider={result['provider']} model={result['model']} dim={len(vectors[0])}")

    # Compare the first text against the others by cosine similarity.
    for other, text in zip(vectors[1:], texts[1:]):
        print(f"  sim({texts[0]!r}, {text!r}) = {cosine(vectors[0], other):.4f}")


if __name__ == "__main__":
    main()
